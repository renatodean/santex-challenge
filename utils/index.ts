import { Item } from "../types/product";

/**
 * From a given product it will return the price as a formated string
 * @param product
 * @param quantity
 * @returns string
 */
export const getPrice = (product: Item, quantity?: number) => {
  const { variants } = product || {};
  const productPrice = variants && variants.length > 0 ? variants[0].price : 0;

  return ((productPrice * (quantity || 1)) / 100).toFixed(2);
};

/**
 * From a given product it will return the price as number
 * @param product
 * @param quantity
 * @returns number
 */
export const getPriceNumber = (product: Item, quantity?: number) => {
  const { variants } = product || {};
  const productPrice = variants && variants.length > 0 ? variants[0].price : 0;

  return (productPrice * (quantity || 1)) / 100;
};

/**
 * From a given query from router it will return the query value
 * @param query
 * @returns string
 */
export const getQueryValue = (query: any) => {
  if (Array.isArray(query)) {
    return query[0];
  } else {
    return query || "";
  }
};
