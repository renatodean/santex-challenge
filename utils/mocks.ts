export const mockedProducts = {
  products: {
    items: [
      {
        __typename: "Product",
        id: "1",
        name: "Laptop",
        slug: "laptop",
        description:
          "Now equipped with seventh-generation Intel Core processors, Laptop is snappier than ever. From daily tasks like launching apps and opening files to more advanced computing, you can power through your day thanks to faster SSDs and Turbo Boost processing up to 3.6GHz.",
        featuredAsset: {
          __typename: "Asset",
          id: "1",
          name: "derick-david-409858-unsplash.jpg",
          type: "IMAGE",
          source:
            "https://demo.vendure.io/assets/source/b6/derick-david-409858-unsplash.jpg",
          preview:
            "https://demo.vendure.io/assets/preview/71/derick-david-409858-unsplash__preview.jpg",
        },
        variants: [
          { __typename: "ProductVariant", price: 129900 },
          { __typename: "ProductVariant", price: 139900 },
          { __typename: "ProductVariant", price: 219900 },
          { __typename: "ProductVariant", price: 229900 },
        ],
      },
      {
        __typename: "Product",
        id: "2",
        name: "Tablet",
        slug: "tablet",
        description:
          'If the computer were invented today, what would it look like? It would be powerful enough for any task. So mobile you could take it everywhere. And so intuitive you could use it any way you wanted — with touch, a keyboard, or even a pencil. In other words, it wouldn’t really be a "computer." It would be Tablet.',
        featuredAsset: {
          __typename: "Asset",
          id: "2",
          name: "kelly-sikkema-685291-unsplash.jpg",
          type: "IMAGE",
          source:
            "https://demo.vendure.io/assets/source/5a/kelly-sikkema-685291-unsplash.jpg",
          preview:
            "https://demo.vendure.io/assets/preview/b8/kelly-sikkema-685291-unsplash__preview.jpg",
        },
        variants: [
          { __typename: "ProductVariant", price: 32900 },
          { __typename: "ProductVariant", price: 44500 },
        ],
      },
    ],
  },
};

export const mockedEmptyProducts = { products: { items: [] } };

export const mockedProduct = {
  product: {
    __typename: "Product",
    id: "1",
    name: "Laptop",
    slug: "laptop",
    description:
      "Now equipped with seventh-generation Intel Core processors, Laptop is snappier than ever. From daily tasks like launching apps and opening files to more advanced computing, you can power through your day thanks to faster SSDs and Turbo Boost processing up to 3.6GHz.",
    featuredAsset: {
      __typename: "Asset",
      id: "1",
      name: "derick-david-409858-unsplash.jpg",
      type: "IMAGE",
      source:
        "https://demo.vendure.io/assets/source/b6/derick-david-409858-unsplash.jpg",
      preview:
        "https://demo.vendure.io/assets/preview/71/derick-david-409858-unsplash__preview.jpg",
    },
    variants: [
      { __typename: "ProductVariant", price: 129900 },
      { __typename: "ProductVariant", price: 139900 },
      { __typename: "ProductVariant", price: 219900 },
      { __typename: "ProductVariant", price: 229900 },
    ],
  },
};

export const mockedStorageState = {
  showSnackbar: true,
  products: [
    {
      product: mockedProduct.product,
      quantity: 3,
    },
  ],
};
