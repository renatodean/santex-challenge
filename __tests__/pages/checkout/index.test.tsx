import { render, screen } from "@testing-library/react";
import React from "react";
import Checkout from "../../../pages/checkout";
import { mockedProduct, mockedStorageState } from "../../../utils/mocks";

const routerPushSpy = jest.fn();

jest.mock("next/router", () => ({
  useRouter() {
    return {
      route: "/",
      pathname: "",
      query: "",
      asPath: "",
      push: routerPushSpy,
      events: {
        on: jest.fn(),
        off: jest.fn(),
      },
      beforePopState: jest.fn(() => null),
      prefetch: jest.fn(() => null),
    };
  },
}));

jest.mock("@apollo/client", () => ({
  __esModule: true,
  useQuery: jest.fn(() => ({ data: mockedProduct })),
  gql: jest.fn(() => ({})),
}));

jest.mock("../../../context", () => ({
  __esModule: true,
  useStorageContext: jest.fn(() => [mockedStorageState, jest.fn()]),
}));

describe("Product", () => {
  it("renders a heading", () => {
    render(<Checkout />);

    const heading = screen.getByRole("heading", {
      name: "Checkout information",
    });

    expect(heading).toBeInTheDocument();
  });

  it("renders one product with quantity 3", () => {
    render(<Checkout />);

    const product = screen.getByText("3x Laptop [$1299.00]");

    expect(product).toBeInTheDocument();
  });
});

describe("Product snapshot", () => {
  it("renders products", () => {
    const { container } = render(<Checkout />);
    expect(container).toMatchSnapshot();
  });
});
