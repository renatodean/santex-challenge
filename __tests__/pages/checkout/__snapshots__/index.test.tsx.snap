// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product snapshot renders products 1`] = `
<div>
  <div
    class="css-1pq8ggb"
  >
    <header
      class="css-1wnfhlx"
    >
      <div
        class="css-1ui6qe5"
        data-testid="home-link"
        tabindex="0"
      >
        <span
          style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
        >
          <img
            alt="Santex Logo"
            data-nimg="fill"
            decoding="async"
            sizes="100vw"
            src="/_next/image?url=%2Fsantex-logo.png&w=3840&q=75"
            srcset="/_next/image?url=%2Fsantex-logo.png&w=640&q=75 640w, /_next/image?url=%2Fsantex-logo.png&w=750&q=75 750w, /_next/image?url=%2Fsantex-logo.png&w=828&q=75 828w, /_next/image?url=%2Fsantex-logo.png&w=1080&q=75 1080w, /_next/image?url=%2Fsantex-logo.png&w=1200&q=75 1200w, /_next/image?url=%2Fsantex-logo.png&w=1920&q=75 1920w, /_next/image?url=%2Fsantex-logo.png&w=2048&q=75 2048w, /_next/image?url=%2Fsantex-logo.png&w=3840&q=75 3840w"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
        </span>
      </div>
      <div
        class="css-1jdhora"
      >
        <div
          class="css-piilf0"
          data-testid="cart-icon"
          tabindex="0"
        >
          <svg
            aria-hidden="true"
            class="StyledIconBase-ea9ulj-0 iLHuDN"
            fill="currentColor"
            focusable="false"
            height="30"
            viewBox="0 0 16 16"
            width="30"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 1.5A.5.5 0 0 1 .5 1H2a.5.5 0 0 1 .485.379L2.89 3H14.5a.5.5 0 0 1 .491.592l-1.5 8A.5.5 0 0 1 13 12H4a.5.5 0 0 1-.491-.408L2.01 3.607 1.61 2H.5a.5.5 0 0 1-.5-.5zM5 12a2 2 0 1 0 0 4 2 2 0 0 0 0-4zm7 0a2 2 0 1 0 0 4 2 2 0 0 0 0-4zm-7 1a1 1 0 1 1 0 2 1 1 0 0 1 0-2zm7 0a1 1 0 1 1 0 2 1 1 0 0 1 0-2z"
            />
          </svg>
          <div
            class="css-1d0bl5n"
          >
            3
          </div>
        </div>
      </div>
    </header>
    <main
      class="css-cf9bqn"
    >
      <h1
        class="css-1yltisf"
      >
        Checkout information
      </h1>
      <div
        class="css-1l2tmoc"
      >
        <div
          class="css-wbbjgv"
        >
          <h3
            class="css-1v2zqs"
          >
            Products
          </h3>
          <div
            class="css-13y32wl"
          >
            <div
              class="css-17ievcd"
            >
              <span
                style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
              >
                <img
                  alt="Laptop"
                  class="css-4z07o0"
                  data-nimg="fill"
                  decoding="async"
                  sizes="100vw"
                  src="/_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=3840&q=75"
                  srcset="/_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=3840&q=75 3840w"
                  style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                />
              </span>
            </div>
            <div
              class="css-rq35md"
            >
              <div
                class="css-e165ny"
              >
                3x 
                Laptop
                 [$1299.00]
              </div>
              <div
                class="css-1ym2u4i"
              >
                $ 
                3897.00
              </div>
            </div>
          </div>
        </div>
        <div
          class="css-ckvj9p"
        >
          <h3
            class="css-1v2zqs"
          >
            Total $ 
            3897.00
          </h3>
          <button
            class="css-z339wo"
          >
            Pay
          </button>
        </div>
      </div>
    </main>
    <header
      class="css-1cvg25p"
    >
      Santex Challenge - Ecommerce
    </header>
  </div>
</div>
`;
